### PR TITLE
Finalizing Direct Offer contract

### PR DESCRIPTION
**Pull request by nikhils9**
- Rename direct-order to direct-offer as per Catalyst proposal project name
- Add fix and property based testing for puniqueOrdered
- Add some more unit tests
- Add detailed user documentation

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,8 +1,166 @@
...
+
+This project is funded by the Cardano Treasury in [Catalyst Fund 10](https://projectcatalyst.io/funds/10/f10-developer-ecosystem-the-evolution/plug-and-play-smart-contract-api-a-game-changing-platform-to-deploy-open-source-contracts-instantly)
+
+## Documentation
+
+### What is Peer-to-Peer (P2P) trading?
+
+P2P trading in the context of this project refers to the direct buying and selling of Cardano Native Tokens (both Fungible & Non-Fungible Tokens) among users, without a third party or an intermediary. This is unlike buying and selling digital assets using a Centralized Exchange (CEX), where you cannot transact directly with counterparties or a DEX where you trade against a fixed Liquidity Pool.
+
+A CEX would use charts and market order aggregators to gauge the current market prices and determine the optimal time to buy, sell, or hold your assets. When you are ready to buy or sell, the exchange enters your order into its order book and facilitates the transaction on your behalf.
```
Comment (mpetruska) [inline README.md]: Well, the first sentence is not really true, a centralized exchange essentially only facilitates an order book and takes the fees on executed trades.

Comment (mpetruska) [inline README.md]:
```suggestion
Trading on a CEX requires you to give custody of your tokens to them, so they can execute the trades you enter based on their charts and market order aggregators. A CEX provides access to their order book and facilitates trades and takes fees in exchange.
```

File: README.md [hunk 2]
```diff
@@ -1,8 +1,166 @@
...
+
+Depending on the type of order you use, effects such as slippage may mean you don’t get the exact price you want. P2P trading, on the other hand, gives you full control over pricing, settlement time, and whom you choose to sell to and buy from.
```
Comment (mpetruska) [inline README.md]:
```suggestion
Depending on the type of order you use, effects such as slippage may mean you don’t get the exact price you want. P2P trading, on the other hand, gives you full control over pricing, settlement time, and whom you choose to sell to and buy from. What is even better, during the whole operation there's no need to give up custody of your assets, no need to trust a centralized entity.
```

File: README.md [hunk 3]
```diff
@@ -1,8 +1,166 @@
...
+
+### How can this project facilitate P2P trading?
+
+This project fulfills the conerstone requirement of a trusted Escrow, over seeing the trade in the form of a smart contract. It locks the seller's assets in it until a buyer provides the required ask or the seller wishes to cancel the offer and claim the funds back.
```
Comment (mpetruska) [inline README.md]:
```suggestion
This project fulfills the cornerstone requirement of a trusted Escrow, over seeing the trade in the form of a smart contract. It locks the seller's assets in the contract until a buyer provides the required ask price or the seller wishes to cancel the offer and claim the funds back.
```

File: README.md [hunk 4]
```diff
@@ -1,8 +1,166 @@
...
+
+### Details
+
+#### Basics of a single offer
+
+A seller sends the assets he wishes to exchange, to the contract, thus locking the funds therein. This UTxO contains a datum which specifies the "creator" of the offer and the assets wished in return, "toBuy" the offered assets.
```
Comment (mpetruska) [inline README.md]:
```suggestion
A seller sends the assets he wishes to exchange, to the contract, thus locking the funds therein. This UTxO contains a datum which specifies the "creator" of the offer and the assets wished in return, "toBuy" indicates the price.
```

File: README.md [hunk 5]
```diff
@@ -1,8 +1,166 @@
...
+          s
+          ( PDataRecord
+              '[ "creator" ':= PAddress
+               , "toBuy" ':= PValue 'Sorted 'Positive
+               ]
+          )
+      )
+```
+
+A buyer interested in the offered assets initiates a transaction (with a "PExecuteOffer" redeemer) spending the locked UTxO by sending it to himself, along with sending the "toBuy" assets to the "creator". This condition is validated by the contract first before allowing the transaction to proceed further. However, if the seller decides to cancel the offer he can do so by initiating a transaction (with a "PReclaim" redeemer) and claim all the locked assets back.
```
Comment (mpetruska) [inline README.md]:
```suggestion
A buyer interested in the offered assets initiates a transaction (with a "PExecuteOffer" redeemer) spending the locked UTxO, along with sending the "toBuy" assets to the "creator". This condition is validated by the contract first before allowing the transaction to proceed further. However, if the seller decides to cancel the offer, he can do so by initiating a transaction (with a "PReclaim" redeemer) and claim all the locked assets back.
```

**Review by mpetruska — COMMENTED** [review]
:+1:, very nice job. As for the documentation: take my suggestions with a grain of salt, English is my second language.

File: README.md (hunk 2 above)
Comment (nikhils9) [reply README.md]: But we would be giving up custody when we lock the assets in the contract and would still have to trust the contract logic.

Comment (mpetruska) [reply README.md]: You are right... how about this one?
```suggestion
Depending on the type of order you use, effects such as slippage may mean you don’t get the exact price you want. P2P trading, on the other hand, gives you full control over pricing, settlement time, and whom you choose to sell to and buy from. What is even better you don't need to give custody of your assets to a centralized entity, they are locked in a contract from which you can reclaim them up until the point they are bought.
```

**Review by mpetruska — APPROVED** [review]

**Pull request merged**